platform/wayland/cog-platform-wl.c

```diff
@@ -1335,24 +1357,19 @@ cog_wl_platform_setup(CogPlatform *platform, CogShell *shell G_GNUC_UNUSED, cons
         return FALSE;
     }
 
-    if (!cog_wl_viewport_create_window(COG_WL_VIEWPORT(self->viewport), error)) {
-        clear_egl(display);
-        clear_wayland(COG_WL_PLATFORM(platform));
-        return FALSE;
-    }
-
     /* init WPE host data */
     wpe_fdo_initialize_for_egl_display(display->egl_display);
 
+    /* get a reference to the hashmap with viewports added to the shell */
+    self->viewports = cog_shell_get_viewports(shell);
+
+    g_signal_connect_swapped(shell, "create-viewport", G_CALLBACK(cog_wl_platform_on_create_viewport), self);
```

Comment:
This allows the specific platforms to react to it and run specific steps. Like here https://github.com/Igalia/cog/pull/675/files#diff-7342248ee4ad822cd294d9c8341a52fd10b3d1df4dd2a202ccb926df67d51034R1327  where the WL platform creates the CogWlWindow.